crates/sui-core/src/node_sync/node_follower.rs

```diff
@@ -358,66 +379,111 @@ where
+                            .committee
+                            .names()
+                            .filter(|n| **n != self.state.name)
+                            .cloned()
+                            .collect();
+                        (None, Some(authorities))
+                    }
+                }
             }
         };
 
-        // Download the cert and effects now that we have established finality and we know that the
-        // effects digest is correct.
-        let (cert, effects) = self
-            .download_cert_and_effects(authorities_with_cert, &digests)
-            .await?;
+        // TODO: perhaps we should start storing these things in Box<>s.
+        #[allow(clippy::large_enum_variant)]
+        enum CertAndEffects {
+            Fullnode(CertifiedTransaction, SignedTransactionEffects),
+            Validator(CertifiedTransaction),
+        }
+
+        let cert_and_effects = match (self.state.is_fullnode(), cert) {
+            (false, Some(cert)) => CertAndEffects::Validator(cert),
+            (is_fullnode, _) => {
+                // Download the cert and effects - either finality has been establish (above), or
+                // we are a validator.
+                let (cert, effects) = self
+                    .download_cert_and_effects(authorities_with_cert, digest, effects_digest)
+                    .await?;
+
+                if is_fullnode {
+                    CertAndEffects::Fullnode(cert, effects)
+                } else {
+                    CertAndEffects::Validator(cert)
+                }
+            }
+        };
 
         // we're done downloading at this point, so we no longer need to prevent other tasks from
         // starting.
         std::mem::drop(permit);
 
-        for parent in effects.effects.dependencies.iter() {
-            let (_, mut rx) = self.pending_txes.wait(parent).await;
+        match cert_and_effects {
+            CertAndEffects::Fullnode(cert, effects) => {
+                for parent in effects.effects.dependencies.iter() {
+                    let (_, mut rx) = self.pending_txes.wait(parent).await;
 
-            if self.state.database.effects_exists(parent)? {
-                continue;
-            }
+                    if self.state.database.effects_exists(parent)? {
+                        continue;
+                    }
 
-            trace!(?parent, digest = ?digests.transaction, "waiting for parent");
-            // Since we no longer hold the semaphore permit, can be sure that our parent will be
-            // able to start.
-            rx.recv()
-                .await
-                .map_err(|e| SuiError::GenericAuthorityError {
-                    error: format!("{:?}", e),
-                })?;
-        }
+                    trace!(?parent, ?digest, "waiting for parent");
+                    // Since we no longer hold the semaphore permit, can be sure that our parent will be
```

Comment:
Not sure this is good enough to prevent a deadlock, what if something else takes this permit and also blocks? Is that possible?

Reply:
If you read from the top of `process_digest`, you'll see that the only blocking task we do while holding the permit is downloading the cert and effects, after which we drop the permit. Downloading will always complete (successfully or otherwise), so permits cannot be held indefinitely.